src/Compilers/Core/Portable/Diagnostic/Diagnostic.DiagnosticWithProgrammaticSuppression.cs

```diff
@@ -0,0 +1,143 @@
+        {
+            private readonly Diagnostic _originalUnsuppressedDiagnostic;
+            private readonly ProgrammaticSuppressionInfo _programmaticSuppressionInfo;
+
+            public DiagnosticWithProgrammaticSuppression(
+                Diagnostic originalUnsuppressedDiagnostic,
+                ProgrammaticSuppressionInfo programmaticSuppressionInfo)
+            {
+                Debug.Assert(!originalUnsuppressedDiagnostic.IsSuppressed);
+                Debug.Assert(originalUnsuppressedDiagnostic.ProgrammaticSuppressionInfo == null);
+                Debug.Assert(programmaticSuppressionInfo != null);
+
+                _originalUnsuppressedDiagnostic = originalUnsuppressedDiagnostic;
+                _programmaticSuppressionInfo = programmaticSuppressionInfo;
+            }
+
+            public override DiagnosticDescriptor Descriptor
+            {
+                get { return _originalUnsuppressedDiagnostic.Descriptor; }
+            }
+
+            public override string Id
+            {
+                get { return Descriptor.Id; }
+            }
+
+            public override string GetMessage(IFormatProvider formatProvider = null)
+                => _originalUnsuppressedDiagnostic.GetMessage(formatProvider);
+
+            internal override IReadOnlyList<object> Arguments
+            {
+                get { return _originalUnsuppressedDiagnostic.Arguments; }
+            }
+
+            public override DiagnosticSeverity Severity
+            {
+                get { return _originalUnsuppressedDiagnostic.Severity; }
+            }
+
+            public override bool IsSuppressed
+            {
+                get { return true; }
+            }
+
+            internal override ProgrammaticSuppressionInfo ProgrammaticSuppressionInfo
+            {
+                get { return _programmaticSuppressionInfo; }
+            }
+
+            public override int WarningLevel
+            {
+                get { return _originalUnsuppressedDiagnostic.WarningLevel; }
+            }
+
+            public override Location Location
+            {
+                get { return _originalUnsuppressedDiagnostic.Location; }
+            }
+
+            public override IReadOnlyList<Location> AdditionalLocations
+            {
+                get { return _originalUnsuppressedDiagnostic.AdditionalLocations; }
+            }
+
+            public override ImmutableDictionary<string, string> Properties
+            {
+                get { return _originalUnsuppressedDiagnostic.Properties; }
+            }
+
+            public override bool Equals(Diagnostic obj)
+            {
+                var other = obj as DiagnosticWithProgrammaticSuppression;
+                if (other == null)
+                {
+                    return false;
+                }
+
+                return Equals(_originalUnsuppressedDiagnostic, other._originalUnsuppressedDiagnostic) &&
```

Comment:
>return [](start = 16, length = 6)

Consider checking if this is the same instance before doing more expensive comparison.